### PR TITLE
Draft proposal: address several bugs; refactor to TypeScript

### PR DESCRIPTION
**Pull request by rossng**
Hi,

I have been adapting the `rehype-twemojify` code to solve some issues that I encountered while adopting it in [our codebase](https://github.com/clowdr-app/clowdr/).

There were two main business logic issues I encountered:

* The grapheme splitting regex doesn't appear to handle some of the newer emoji correctly. I switched to using the `emoji-regex` library for everything, inspired by the approach of the `react-emoji-render` library.
* Twemoji strips the trailing uFE0F from the code point of non-emoji graphemes that are hinted into being emoji (like ☘️). See for example the [logic used in the `twemoji` library](https://github.com/twitter/twemoji/blob/793a6a93f303c08877dd6eb589b2fabb3d1c45ee/scripts/build.js#L346).

Because I'm most comfortable working in TypeScript+pnpm, I ended up doing some more extensive refactoring to solve various issues I found with inconsistent types in the codebase. This includes various backwards-incompatible changes.

This PR also depends on some as-yet-unmerged [fixes I made to the `unist-util-map` types](https://github.com/syntax-tree/unist-util-map/pull/3).

I appreciate that these other changes might take this PR past the point where you actually want to merge it into the existing codebase. However, I'd be happy to work with you to find a set of changes that you would want to merge. (I didn't just want to run off and fork the codebase needlessy!)

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    };
+  };
+}
+
+/**
+ * Plugin to twemoji-fy ordinary emojis in HTML.
+ */
+const rehypeTwemojify: Plugin<[UserOptions?], Root, Root> = function (userOptions) {
+  const options = resolveOptions(userOptions);
+  return makeTranformer(options);
```
Comment (cliid) [inline src/index.ts:134]: Seems to be a typo.

Comment (rossng) [reply src/index.ts]: Well-spotted! I've fixed that now.

**Comment by rossng** [comment]
I came to the conclusion that it's very hard to fix the types of `unist-util-map`, so I have resorted to a couple of type casts instead.

**Comment by cliid** [comment]
Thanks for contributing, rossng!

**Pull request merged**